gprofiler/python.py

```diff
@@ -204,39 +228,47 @@ def start(self):
             # Duration is irrelevant here, we want to run continuously.
         ]
         process = start_process(cmd)
+        # wait until the transient data file appears - because once returning from here, PyPerf may
+        # be polled via snapshot() and we need it to finish installing its signal handler.
         try:
-            poll_process(process, self.poll_timeout, self._stop_event)
+            wait_event(self.poll_timeout, self._stop_event, lambda: os.path.exists(self.output_path))
         except TimeoutError:
-            # Process is alive. So far, so good...
-            if not self.output_path.exists():
-                process.kill()
-                stdout = process.stdout.read().decode()
-                stderr = process.stderr.read().decode()
-                self._check_missing_headers(stdout)
-                raise CalledProcessError(process.returncode, process.args, stdout, stderr)
-            # happy flow
-            self.process = process
+            process.kill()
+            raise
         else:
-            # process terminated, must be an error:
-            self._check_output(process, self.output_path)
-            raise RuntimeError("_check_output did not raise!")
+            self.process = process
 
-    def _dump(self) -> Path:
-        assert self.process is not None, "profiling not started!"
-        self.process.send_signal(self.dump_signal)
+    def _glob_output(self) -> List[str]:
         # important to not grab the transient data file
-        while True:
-            output_files = glob.glob(f"{str(self.output_path)}.*")
-            if output_files:
-                break
+        return glob.glob(f"{str(self.output_path)}.*")
 
-            if self._stop_event.wait(0.1):
-                raise StopEventSetException()
+    def _wait_for_output_file(self, timeout: float) -> Path:
+        wait_event(timeout, self._stop_event, lambda: len(self._glob_output()) > 0)
 
+        output_files = self._glob_output()
         # All the snapshot samples should be in one file
         assert len(output_files) == 1
         return Path(output_files[0])
 
+    def _dump(self) -> Path:
+        assert self.process is not None, "profiling not started!"
+        self.process.send_signal(self.dump_signal)
+
+        try:
+            return self._wait_for_output_file(self.dump_timeout)
+        except TimeoutError:
+            # error flow :(
+            if _reinitialize_profiler is not None:
+                logger.warning("Reverting to py-spy")
+                global _profiler_class
+                _profiler_class = PySpyProfiler
+                _reinitialize_profiler()
```

Comment:
`_pyperf_error` raises, and this may raise as well, so to make this safe I think the easiest option is to place one in a `try` and the other in a `finally`. Doesn't matter which comes first, I think.

Reply:
Can't really raise now (since py-spy c'tor does nothing). That's also why I put it first.
But agree that the 2 are unrelated and should happen regardless the possible failure of the others.